From 015cde7dc2c54c63ab959b191e369b47440fa93d Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Sat, 21 May 2022 23:08:09 +0200
Subject: [PATCH] Use old device tree on altboot

Variscite renamed device trees starting with
the kernel 5.4.85 release, let's load the old
device tree name in case of a rollback.

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 include/configs/imx8mm_var_dart.h | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/include/configs/imx8mm_var_dart.h b/include/configs/imx8mm_var_dart.h
index b722fba0e9..dd13f40105 100644
--- a/include/configs/imx8mm_var_dart.h
+++ b/include/configs/imx8mm_var_dart.h
@@ -132,9 +132,15 @@
 				"fi; " \
 			"fi; " \
 		"fi; \0" \
+	"imx_boot_bin=imx-boot-imx8mm-var-dart-plt-sd.bin-flash_lpddr4_ddr4_evk\0" \
+	"set_num_blocks=setexpr imx_boot_num_blocks ${filesize} / 0x200 ; setexpr imx_boot_num_blocks ${imx_boot_num_blocks} + 1; echo imx-boot has ${imx_boot_num_blocks} blocks; \0" \
+	"revert_imx_boot=load mmc ${mmcdev}:${resin_root_part} ${loadaddr} boot/${imx_boot_bin} ; " \
+                       "run set_num_blocks; mmc write ${loadaddr} 0x42 ${imx_boot_num_blocks}; reset; \0" \
 	"loadfdt=run findfdt; " \
-		"echo fdt_file=${fdt_file}; " \
-		"load mmc ${mmcdev}:${resin_root_part} ${fdt_addr} boot/${fdt_file}\0" \
+		"echo fdt_file=${fdt_file}; setenv old_fdt imx8mm-var-dart-legacy.dtb; " \
+		"if load mmc ${mmcdev}:${resin_root_part} ${fdt_addr} boot/${fdt_file}; then echo Loaded ${fdt_file}; " \
+		"else if load mmc ${mmcdev}:${resin_root_part} ${fdt_addr} boot/${old_fdt}; then echo Using old legacy dt!; " \
+		"else if test ${mmcdev} = 2; then echo Will revert imx-boot for compatibility; run revert_imx_boot; fi; fi; fi; \0" \
 	"ramsize_check="\
 		"if test $sdram_size -le 512; then " \
 			"setenv cma_size cma=320M; " \
-- 
2.17.1

